### PR TITLE
Fix 9523 - real estate (like castles) is not automatically taken over by a new territory owner.

### PR DESCRIPTION
**Pull request by RaiNova**
## Change Summary & Additional Notes

As issue #9523 describes, real estate like castles is not automatically taken over by a new territory owner.
The PR fixes that. It changes the ownership for all units on the respective territory of unit attachments with the following characteristics 
- the unit cannot be transported by land
- the unit cannot be transported by air
- being owned by the old owner the unit has no movement points

### CODE REVIEW

**Comment by RaiNova** [comment]
@DanVanAtta I though I'd just commit two files here (one main file and one test file). Now it's the whole shebang again. What can I do to tidy up?

**Comment by RaiNova** [comment]
If this is merged with the current head, then only the two relevant files should remain. Can _I_ do that? If not: Can you reject the PR, and I submit it anew - that may boild it down also.

**Comment by RoiEXLab** [comment]
@RaiNova Basically you have to rebase the branch `fix_9523` so that it starts off `triplea-game/master`. (Technical term 'rebase fix_9523 onto master'). Currently the branch and therefore this PR diverges from the master branch in the past not the present. This is why the whole diff is displayed instead of just the 2 most recent commits.
If you have any issues trying to rebase your branch, feel free to ask and I might be able to help, but I won't teach you git from the ground up.

**Comment by DanVanAtta** [comment]
@RaiNova , FWIW I've a repo with git commands that I run. Generally I've found that wrapping any common/frequent git commands is really useful. That and using [scm breeze](https://github.com/scmbreeze/scm_breeze).

In the repo, you would be looking for the `gup` command, it is short for a `git pull --rebase` (and `gup` does some fun color coding for clarity).

Second, when doing development, at least for this project, one should generally prefer to always create a new branch and submit that branch for PR. The 'master' branch in your fork would then be kept to track with the master branch on origin, and you would update your 'master' branch frequently to pull in any new updates. From there you can then create new branches that are even with master and submit those for PR.

 If it helps, the typical git/PR workflow has some notes here: https://github.com/triplea-game/triplea/blob/master/docs/development/how-to/typical-git-workflow.md

**Comment by Cernelius** [comment]
Would this apply also to non-infrastructure units? For example, when a territory is turned over to an ally in Napoleonic Empires, will all mortella towers and fortresses in the territory change ownership the same way?

I'm thinking it may be better just to have the inability to move as a condition, ignoring any inabilities to being transported.

**Comment by RaiNova** [comment]
>@DanVanAtta  If it helps, the typical git/PR workflow has some notes here: https://github.com/triplea-game/triplea/blob/master/docs/development/how-to/typical-git-workflow.md

Maybe I do something fundamentaly different. I have created my own fork of `tiplea-game:master` to work on.
That worked for my first PR. Is that _not_ the right way to go about?

Directly working on `tiplea-game:master`  does _not_ work for my but leads to the followin error message:

```
21:54:00.535: [tripleaOriginal] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false add --ignore-errors -A -f -- game-app/game-core/src/main/java/games/strategy/engine/data/changefactory/OwnerChange.java
21:54:00.875: [tripleaOriginal] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false commit -F C:\Users\rnova_ebav3ic\AppData\Local\Temp\git-commit-msg-.txt --
[master 19e097823] Fix 9523 again, this time not from my fork
 1 file changed, 14 insertions(+), 1 deletion(-)
21:54:04.741: [tripleaOriginal] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false push --progress --porcelain origin refs/heads/master:master
remote: Permission to triplea-game/triplea.git denied to RaiNova.
fatal: unable to access 'https://github.com/triplea-game/triplea/': The requested URL returned error: 403
21:54:05.904: [tripleaOriginal] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false push --progress --porcelain origin refs/heads/master:master
remote: Permission to triplea-game/triplea.git denied to RaiNova.
fatal: unable to access 'https://github.com/triplea-game/triplea/': The requested URL returned error: 403

```

**Comment by RaiNova** [comment]
This is how my fork looks inside IntelliJ
image

**Comment by RaiNova** [comment]
I also tried from www.github.com:
image

image
...
image

Do I understand correctly, that I cannot use the above command line steps, because tribalea-game:master is protected?

**Comment by RaiNova** [comment]
Next try with Github Desktop:
image

Is this the way to go? Why does my master have 37 commits (which commits?)? What effect would it have to apply them "on top of upstream/master"? Is `upstream/master` `triplea-game:master`?

"This will update master by appying 37 commits of upstream/master on top of master" would make sense to me. Is this the intended meaning of the sentence?

... I started the rebase anyway. I took me about 1h to resolve some 100 conflicts which were always different versions of spacing and line breaking. Was there an option to say "I always accept the HEAD version" that I missed?

Now github shows my master is not behind any more:
image

But my fix 9523 branch still is:
image

How can I apply the same rebase / conflict resolution that I spent an hour on for the master apply to the fix branch?

**Comment by RoiEXLab** [comment]
@RaiNova Ok some things to clarify:
From what I understand you rebased your branch `RaiNova:master` onto `triple-game:master`, i.e. make `RaiNova:master` diverge from `triple-game:master` after the latest commit. Does your master branch have any of the changes you wanted to implement in this PR? If yes, you can just open a PR that tries to merge your master branch into triplea-game's master branch. If not then chances are your effort was for nothing unfortunately.
This PR tracks your `fix_9523` branch and all changes on it, so in order to affect this PR in any way you need to change this branch and this branch only (not your master branch, which is completely separate). This can happen via a merge commit, which merges 2 branches, or via a rebase that changes the history of your branch as if your branch was created now instead of the past.

Have a look at the following images from the git history:
Git Start
Git end

The light blue line represents the branch of your Pull Request, the dark blue line the master branch of triplea-game. As you can see the commit where both diverge is "Consolidate 'run_ansible' scripts (#9507)" from August last year, which explains all of the commits you can see here. Also note how merge commits only keep your branch "up-to-date", but doesn't change the commit where the branches branch off each other. A rebase would do that, rewrite history and move the diverging commit from the past into the present.

So to get your master branch even with triplea-game's master branch you'd either have to rebase it every now and then without making any commits to it whatsoever (which doesn't really work anymore because a rebase doesn't undo any commits you made to this branch) or HARD reset your branch to `triplea-game:master` and force push that branch. Note that this will permanently delete any commits on this branch that cannot be found on `triplea-game:master` without any warning, so handle with care.


Regarding your question:

> Do I understand correctly, that I cannot use the above command line steps, because tribalea-game:master is protected?

Yes and no. Yes in the sense that nobody can push to master because of its protection. And no in the sense that only some members of the `triplea-game` org have write access to the repository. Otherwise anyone could just make a commit to master without anyone noticing, this system enforces the PR review system which is a good thing, manual merging is something we don't ever want to do unless something is inheritely broken.

So how should you proceed?
I can't give you a definitive answer, but my personal recommendation would be to try to merge `triplea-game:master` into `RaiNova:fix_9523` and push the changes (can also be done in the browser I think, if command line or GitHub Desktop is too complicated for you) so we can merge them.

Once it's merged and your work is safely integrated you can try to HARD reset your master branch to `triple-game:master` and create any future branches based off `triplea-game:master` to get a clean fresh working copy for any future PRs.

It's perhaps worth mentioning that your local master branch is in theory completely separate from your online master branch. Pushing and pulling is what keeps them in sync.
Also to match the terminology of GitHub Desktop: `upstream` probably refers to `triple-game` and `origin` to `RaiNova` as alias. Branch names without any prefix are locally.

I hope that helps a little bit

**Comment by RaiNova** [comment]
>@RoiEXLab So how should you proceed?
I can't give you a definitive answer, but my personal recommendation would be to try to merge triplea-game:master into RaiNova:fix_9523 and push the changes (can also be done in the browser I think, if command line or GitHub Desktop is too complicated for you) so we can merge them.

I merged from GitHub Desktop and it worked. Thanks for your help! The PR is now down to 11 files. `OwnerChange.java` is the file the change really is about. `CasualtyDetails.java` contains two more whitespaces the formatter put in.

The other 9 files are - as far as I understand - on an ignore list that GitHub Desktop ignored.

**Comment by RaiNova** [comment]
**Spotless** test result on casualty details: Complains about two white spaces the formatter in IntelliJ inserted. If they are wrong: Could anybody provide a correct formatter please? Otherwise: Could the spotless check please be adjusted?

**Unit tests**: I didn't change anything relating to `ZipExtractorTest ` and it runs fine on my PC.

**Comment by RoiEXLab** [comment]
@RaiNova We use spotless as single source of truth for formatting.
Just run `./gradlew :game-app:game-core:spotlessApply` to auto-format all the files.

Don't worry about the failing unit test, it is expected: See #10050 for more information

**Comment by RaiNova** [comment]
running `gradlew :game-app:game-core:spotlessApply` leads to

```
Starting a Gradle Daemon, 3 busy and 38 incompatible and 1 stopped Daemons could not be reused, use --status for details

> Task :game-app:game-core:spotlessJava FAILED
Step 'google-java-format' found problem in 'src\main\java\games\strategy\engine\auto\update\EngineVersionCheck.java':
null
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$createFormat$1(GoogleJavaFormatStep.java:176)
        at com.diffplug.spotless.FormatterFunc.apply(FormatterFunc.java:32)
        at com.diffplug.spotless.FormatterStepImpl$Standard.format(FormatterStepImpl.java:78)
        at com.diffplug.spotless.FormatterStep$Strict.format(FormatterStep.java:88)
        at com.diffplug.spotless.Formatter.compute(Formatter.java:230)
        at com.diffplug.spotless.PaddedCell.calculateDirtyState(PaddedCell.java:203)
        at com.diffplug.spotless.PaddedCell.calculateDirtyState(PaddedCell.java:190)
        at com.diffplug.gradle.spotless.SpotlessTaskImpl.processInputFile(SpotlessTaskImpl.java:92)
        at com.diffplug.gradle.spotless.SpotlessTaskImpl.performAction(SpotlessTaskImpl.java:78)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.IncrementalInputsTaskAction.doExecute(IncrementalInputsTaskAction.java:32)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
        at org.gradle.api.internal.project.taskfactory.AbstractIncrementalTaskAction.execute(AbstractIncrementalTaskAction.java:25)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$2.run(ExecuteActionsTaskExecuter.java:498)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:483)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:466)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:105)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:270)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:248)
        at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:83)
        at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:37)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:50)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:47)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:37)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:68)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:38)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:50)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:36)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:29)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:54)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:35)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:60)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:27)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:174)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeAndStoreInCache(BuildCacheStep.java:150)
        at org.gradle.internal.execution.steps.BuildCacheStep.lambda$executeWithCache$2(BuildCacheStep.java:124)
        at org.gradle.internal.execution.steps.BuildCacheStep.lambda$executeWithCache$3(BuildCacheStep.java:124)
        at org.gradle.internal.Try$Success.map(Try.java:162)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithCache(BuildCacheStep.java:83)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:73)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:45)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:29)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:36)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:22)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:99)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:92)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:52)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:84)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:41)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:91)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:78)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:49)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:105)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:50)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:86)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:86)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:32)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:43)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:31)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:40)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution$2.withWorkspace(ExecuteActionsTaskExecuter.java:283)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:40)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:30)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:37)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:27)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:49)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:35)
        at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:76)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:408)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:395)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:388)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:374)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.IllegalAccessError: class com.google.googlejavaformat.java.JavaInput (in unnamed module @0x1affe665) cannot access class com.sun.tools.javac.parser.Tokens$TokenKind (in module jdk.compiler) because module jdk.compiler does not export com.sun.tools.javac.parser to unnamed module @0x1affe665
        at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:349)
        at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:334)
        at com.google.googlejavaformat.java.JavaInput.<init>(JavaInput.java:276)
        at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:280)
        at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:267)
        at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:233)
        ... 140 more

FAILURE: Build failed with an exception.
```

**Comment by RaiNova** [comment]
@RoiEXLab , @DanVanAtta Thanks for the help of both of you, which - after hours spent, mainly on whitespacing - empowered me to perform an essentially simple task 😀 

Who volunteers reviewing?

File: game-app/game-core/src/test/java/games/strategy/engine/data/changefactory/OwnerChangeTest.java
Comment (RoiEXLab) [inline game-app/game-core/src/test/java/games/strategy/engine/data/changefactory/OwnerChangeTest.java:18]:
```suggestion
  void testPerform_staticUnitIsTakenOver() {
```

File: game-app/game-core/src/test/java/games/strategy/engine/data/changefactory/OwnerChangeTest.java
Comment (RoiEXLab) [inline game-app/game-core/src/test/java/games/strategy/engine/data/changefactory/OwnerChangeTest.java:40]:
```suggestion
  void testPerform_mobileUnitIsNotTakenOver() {
```

**Review by RoiEXLab — APPROVED** [review]
So the code looks fine from a coding perspective. I have no idea about how the rules are supposed to work here.
I didn't read the whole thread of #9523 but I'll just give you the benefit of the doubt and assume that you discussed this with @Cernelius in detail

**Comment by asvitkine** [comment]
This breaks the various unit attachment properties about what should happen when the unit should be captured, which are implemented in BattleTracker.captureOrDestroyUnits().

**Comment by asvitkine** [comment]
(More specifically, the code in captureOrDestroyUnits() is trying to look for eligible units where the functionality should be applied, but by the time it runs, this new code would have ran and the units are already owned by the new player, so the check for Matches.enemyUnit() excludes the units.)

**Comment by asvitkine** [comment]
Some thoughts:

1. The problem is this logic applies even in the case when the territory was taken over via a battle, but the intention was to just handle the case via changes as a result of Allied political relationship. Can we find a more narrow place to put this logic such that it won't apply for regular battles?
2. If we do the above, I assume we also need to handle the case where there IS a battle, but an ally conquers a territory formally owned by their ally, so we need to check how the two logics interact.
3. Another way to resolve this could be as a new property (so relevant maps would need to be updated), since right now the map in question specifies:
```xml
        <attachment name="relationshipTypeAttachment" attachTo="Allied" javaClass="games.strategy.triplea.attachments.RelationshipTypeAttachment" type="relationship">
            <option name="archeType" value="allied"/>
            <option name="alliancesCanChainTogether" value="true"/>
            <option name="givesBackOriginalTerritories" value="true"/>
            <option name="canMoveThroughCanals" value="true"/>
        </attachment>
```
So it could also specify something like
```xml
            <option name="givesBackInfrastructureUnitsInOriginalTerritories" value="true"/>
```

Then, this new code would be tied in to this new property and maps wishing this behavior could be updated to set it.

**Comment by asvitkine** [comment]
BattleTracker has this code:
```
       // give back the factories as well.
      for (final Territory item : friendlyTerritories) {
        if (item.isOwnedBy(terrOrigOwner)) {
          continue;
        }
        final Change takeOverFriendlyTerritories = ChangeFactory.changeOwner(item, terrOrigOwner);
        addChange(bridge, changeTracker, takeOverFriendlyTerritories);
        bridge.getHistoryWriter().addChildToEvent(takeOverFriendlyTerritories.toString());
        final Collection<Unit> units =
            CollectionUtils.getMatches(item.getUnits(), Matches.unitIsInfrastructure());
        if (!units.isEmpty()) {
          final Change takeOverNonComUnits =
              ChangeFactory.changeOwner(units, terrOrigOwner, territory);
          addChange(bridge, changeTracker, takeOverNonComUnits);
        }
      }
```

So similar code can be added to `PoliticsDelegate.givesBackOriginalTerritories()`

**Comment by asvitkine** [comment]
@RaiNova is this something you can look at fixing in the near future?

If not, I think we should revert this change and re-open the issue it was trying to fix. Then, a different approach at implementing this can be investigated whenever you have time.

**Comment by RaiNova** [comment]
@asvitkine Thanks for the analysis - I'm having a look at it. Could you wirte a test case that currently fails and demonstrates the error?

**Comment by asvitkine** [comment]
I don't have a test case, but the save game attached here can be used to demonstrate the problem:  https://forums.triplea-game.org/topic/2998/2-6-testing-report-2-6-problems-here/68

Land the transported unit in the american territory in italy and the major factory should turn minor. Possibly, you could set something like this up programmatically.

**Comment by asvitkine** [comment]
You could probably set up a test in a similar way to RevisedTest.testRetreatBug - which sets up some board state and moves through steps and resolves combats. You just want to do it on TestMapGameData.GLOBAL1940 instead of revised.

**Comment by TheDog-GH** [comment]
An Edit Mode battle should produce the error, I used it for my test in https://github.com/triplea-game/triplea/issues/10453

**Pull request merged**
